[website]: add databases page to docs (#26562)
ethereum/go-ethereum: docs/fundamentals/databases.md

| `compReadMeter`      | Meter for measuring the data read during compaction                     |
| `compWriteMeter`     | Meter for measuring the data written during compaction                  |
| `writeDelayNMeter`   | Meter for measuring the write delay number due to database compaction   |
| `writeDelayMeter`    | Meter for measuring the write delay duration due to database compaction |
| `diskSizeGauge`      | Gauge for tracking the size of all the levels in the database           |
| `diskReadMeter`      | Meter for measuring the effective amount of data read                   |
| `diskWriteMeter`     | Meter for measuring the effective amount of data written                |
| `memCompGauge`       | Gauge for tracking the number of memory compaction                      |
| `level0CompGauge`    | Gauge for tracking the number of table compaction in level0             |
| `nonlevel0CompGauge` | Gauge for tracking the number of table compaction in non0 level         |
| `seekCompGauge`      | Gauge for tracking the number of table compaction caused by read opt    |

## Freezer/ancients

> Older segments of the chain are moved out of the fast LevelDB database and into a freezer database the is less performant. Nodes rarely need to access these files so IO speed is less important and the bulk of the chain data can be stored on a cheaper HDD. Once blocks pass some threshold age (128 blocks behind the head by default) the block and receipt data is flattened and saved as a raw binary blob of data along with an index entry file used for identification. This flat file format allows Geth to capitalize on the operating system's built-in memory manager and file cacher which in turn allows Geth to maximize the available memory without hogging it from other processes. This is good for overall system stability and also reduces the strain on Go's garbage collector and reduces the total amount of memory used by Geth. 

Reviewer: Once blocks pass some threshold age (90,000 blocks behind the head by default)

Reviewer: > This flat file format allows Geth to capitalize on the operating system's built-in memory manager and file cacher which in turn allows Geth to maximize the available memory without hogging it from other processes

I am not sure about this property. From my own understanding, freezer just opens a batch of files *normally* without any fancy tricks(e.g. mmap).

Reviewer 2: I agree with what is written. We just use the filesystem, and it's up to the OS to manage the filesystem. If the os has memory available, it will use (some of) it as a backing cache for the filesystem. So yea, we let the os do it's thing.

However, I think it's a bit "wordy". I don't see the need to "sell" the flat file format - IMO it's enough to describe it.

Author: ```suggestion
Older segments of the chain are moved out of the fast LevelDB database and into a freezer database the is less performant. Nodes rarely need to access these files so IO speed is less important and the bulk of the chain data can be stored on a cheaper HDD. Once blocks pass some threshold age (90,000 blocks behind the head by default) the block and receipt data is flattened and saved as a raw binary blob of data along with an index entry file used for identification. 
```

Geth also tracks some basic metrics relating to the ancients database that can be monitored:

| metric       | function                                                   |
| ------------ | ---------------------------------------------------------- |
| `readMeter`  | Meter for measuring the effective amount of data read      |
| `writeMeter` | Meter for measuring the effective amount of data written   |
| `sizeGauge`  | Gauge for tracking the combined size of all freezer tables |

> The ancients data is saved entirely separately from the fast-access recent data, meaning it can be stored in a  different location. The default location for the ancient chain segments is inside the `chaindata` directory, which is inside `datadir`, but it can be defined by passing `--datadir.ancient <path>` to Geth on startup. This path can be to a slow HDD disk, permitting people to run Ethereum nodes without requiring a huge SSD. The ancient can also be moved later by manually copying the directory to a new location and then starting Geth passing the new path to `--datadir.ancient`.

Reviewer: I would appreciate if we can advertise our freezer design a bit.

The complexity of read operation in freezer is O(1), which only involves:
- a read for index item(6 bytes)
- a read for data(depends on the size of data)
And that's all. Freezer should still be performant enough on slow HDD
because of this design.

## Using the freezer

> If Geth cannot find the freezer, either because the database was deleted or because Geth received an incorrect path, then the node becomes unstable. It is explicitly forbidden to start Geth with an invalid path to the freezer. However, if the state database goes missing it can be rebuilt "on-the-fly" by Geth. In this case, Geth rebuilds all its indices based upon data from the freezer and then does a snap-sync on top of it to rebuild the missing state data. In summary - Geth can rebuild the recent past from the ancient past but not vice-versa.

Reviewer 2: > However, if the state database goes missing it can be rebuilt "on-the-fly" by Geth.

Wait a minute... How do you mean? You must be thinking of a different scenario than I am, because if you nuke the statedb and keep only ancients, this is not at all possible.

Reviewer: Why? If we plugin the existent ancient into a empty Geth instance, the ancient can be reused and run snap-sync on top.

Author: will wait for clarification from you guys before updating this - my understanding is that losing the state db could be recovered by snap-syncing on top of the ancients. This is also how it is described here in the "Freezer Tricks" section: https://blog.ethereum.org/2019/07/10/geth-v1-9-0.

Reviewer 2: Or rather, you compared two scenarios, and it sounded like you need to be careful with ancients, geth can become unstable if you mess with them. But state can be fixed "on the fly" if you mess it up. 

And that feels a bit odd, to me

Author: ok, maybe just delete the last sentence then? It's kind of unnecessary and might have muddied the waters.

```suggestion
If Geth cannot find the freezer, either because the database was deleted or because Geth received an incorrect path, then the node becomes unstable. It is explicitly forbidden to start Geth with an invalid path to the freezer. However, if the state database goes missing it can be rebuilt "on-the-fly" by Geth. In this case, Geth rebuilds all its indices based upon data from the freezer and then does a snap-sync on top of it to rebuild the missing state data.
```

> This can be used to deliberately clean up a node. Passing `--datadir --removedb` starts the process. Geth will ask whether it should delete the state database, the ancient database and the LES database. Choosing to delete the state database onyl leaves the block bodies, receipts, headers etc intact in the freezer, meaning rebuilding the state will not include re-downloading ~400GB of data from the network. Geth will then rebuild the state from the freezer reusing that existing block and receipt data. In doing so, unused data and accumulated junk data will be pruned from the state database. This process can take an hour or more.

Reviewer: onyl -> only

Author: ```suggestion
This can be used to deliberately clean up a node. Passing `--datadir --removedb` starts the process. Geth will ask whether it should delete the state database, the ancient database and the LES database. Choosing to delete the state database only leaves the block bodies, receipts, headers etc intact in the freezer, meaning rebuilding the state will not include re-downloading ~400GB of data from the network. Geth will then rebuild the state from the freezer reusing that existing block and receipt data. In doing so, unused data and accumulated junk data will be pruned from the state database. This process can take an hour or more.
```

## Unclean shutdowns

> If Geth stops unexpectedly the database can be corrupted. This is known as an "unclean shutdown" and it can lead to a variety of problems for the node when it is restarted. It is always best to shut down Geth gracefully (i.e. using a shutdown command such as `ctrl-c` or `systemctl stop`). This way, Geth knows to write all relevant information into the database to allow the node to restart properly later. This can involve >1GB of information being written to the LevelDB database which can take several minutes.

Reviewer 2: `systemctl stop` -- doesn't that require some timeout parameter?

Author: ah good point- yes it has a short default timeout.
```suggestion
If Geth stops unexpectedly the database can be corrupted. This is known as an "unclean shutdown" and it can lead to a variety of problems for the node when it is restarted. It is always best to shut down Geth gracefully, i.e. using a shutdown command such as `ctrl-c` or `systemctl stop`  (although please note that `systemctl stop` has a default timeout of 90s - if Geth takes longer than this to gracefully shut down it will quit forcefully. Update the `TimeoutSecs` variable in `systemd.service` to override this value to something larger, at least 300s). This way, Geth knows to write all relevant information into the database to allow the node to restart properly later. This can involve >1GB of information being written to the LevelDB database which can take several minutes.
```

> If an unexpected shutdown does occur, the `removedb` flag can be used to delete the state database and resync it from the ancient database. This should get the database back up and running.

Reviewer: Theoretically geth should survive unclean shutdown. It means that for most of the unclean shutdowns, no additional recovery operations required and Geth can fix it by itself.

So I think we need to re-phrase this sentence a bit. e.g. `If a database corruption does occur, ... `

Reviewer 2: nit: `removedb` is not a flag, but a (sub-)command.

Flags are arguments that set details on a command, but the command is the basic verb.

Author: ```suggestion
If an unexpected shutdown does occur, the `removedb` subcommand can be used to delete the state database and resync it from the ancient database. This should get the database back up and running.
```
